### ISSUE TITLE
Logging Remote IP Address

### ISSUE THREAD

**Issue by brycejlowe** [simplesamlphp/simplesamlphp#1151]
I apologize for yet another discussion on load balancers and SimpleSAMLphp, but I do want to get the developers opinion on the appropriate way to accomplish IP address logging in SimpleSAMLphp.

**Please note: I am not advocating for the un-checked use of X-Forwarded-* headers in my comments below, I am simply trying to understand the prescribed way of doing things.**

I do realize that the appropriate way to signal to SimpleSAMLphp that it is behind a load balancer is the use of the configuration variable baseurlpath.  That makes perfect sense to me.

The issue I have is around logging and where the client's IP address is retrieved from (for instance in modules\core\lib\Auth\UserPassBase.php when an unsuccessful login happens, the $_SERVER['REMOTE_ADDR'] variable is queried).  Can someone tell me what the prescribed method for obtaining the client's true IP address is?

In code that I own, I have an array of IP addresses that represent the load balancers in my environment.  If the REMOTE_ADDR is in the list of trusted IPs, I **know** that the load balancer I have configured overrides the X-Forwarded-For header with the client's real ip address, and I **know** that load balancer sits a hop before my SimpleSAMLphp instance, therefore I can trust that the information in X-Forwarded-For is actually the client's IP address and that it wasn't spoofed in the request.  If the REMOTE_ADDR doesn't match the known load balancers, I simply ignore X-Forwarded-For and use REMOTE_ADDR instead.

Given that at least some of the code in SimpleSAMLphp references the raw superglobal $_SERVER['REMOTE_ADDR'], is the only option to overwrite the $_SERVER['REMOTE_ADDR'] with the true IP address of the client by injecting PHP really early in the application lifecycle?  This has a "smell" to me, as I try really hard never to overwrite superglobals as a personal habit.  If I need to modify a superglobal, I usually wrap it in a "request" object (or at least a variable) and use that in my subsequent code rather than the superglobal itself.  In other frameworks I see some type of request object passed around that has the business logic built into it rather than using the raw superglobals themselves.  Can you shed some light on what the right way to do this in SimpleSAMLphp?

**Comment by tvdijen** [comment]
Your load balancer should be perfectly capable of using the real ip towards yoyr backend servers ..

**Comment by brycejlowe** [comment]
Hi @tvdijen,

Thank you for your comments.  However, I am not sure I understand what you are meaning.  The load balance absolutely does provide the real client ip to the backend servers, it does it in the form of a HTTP header X-Forwarded-For.

The actual connection from the external client is terminated at the load balancer (in my case HAProxy), the load balancer then creates a new connection to the SimpleSAMLphp backend on-behalf of the client and the request is sent through (in my case, I modify the request to remove any provided X-Forwarded-For headers from the original HTTP request and HAProxy passes along the real client's ip address using that header).  Because of the new connection from the load balancer, the REMOTE_ADDR is the load balancer's IP address, not the external client's IP address.  What I am asking is what is the recommended way to log the **external** client's IP address in a load balanced (and reverse proxy) scenario?

Thanks,
Bryce

**Comment by tvdijen** [comment]
What you're describing is not load balancing, but reverse proxying...
Doesn't really change anything though.. Instead of passing the client's IP address in the X-FORWARDED-FOR header, it should also be possible to pass it as the REMOTE_ADDR..

**Comment by brycejlowe** [comment]
Hi @tvdijen,

Thank you for your response.  I should have been more clear in my original post that I am indeed using a load balancer that is also a reverse proxy (specifically it is a layer 7 load balancer, not configured to be a transparent reverse proxy).  **Based on your comments, I have concluded that SimpleSAMLphp requires the `$_SERVER['REMOTE_ADDR']` to be the actual address of the client, period**.  I also acknowledge the value of `$_SERVER['REMOTE_ADDR']` for making decisions.  For example I have a custom SimpleSAMLphp module that I wrote that requires 2-factor SMS based authentication for external users that will only trigger the 2 factor if the client's remote IP address is not in a trusted list of RFC1918 addresses.  I actually modified that code to work with my reverse proxy load balancers already, the only missing piece was SimpleSAMLphp logging.

In my environment: SimpleSAMLphp running on Apache via mod_php behind a layer 7 reverse proxy load balancer (not running in transparent mode), it is just a fact that the `$_SERVER['REMOTE_ADDR']` by default contains the reverse proxy load balancer ip address and `X-Forwarded-For` contains the client's IP address.  Luckily for me, I am running Apache 2.4 which has a solution that I am happy with (it doesn't seem like it exists in the default installation of Apache 2.2, although it seems some people have written ones that are compatible with that version) and that is mod_remoteip.

In my Apache configuration (running on CentOS 7) on all backend SimpleSAMLphp servers that are connected to via the reverse proxy I simply did this:

1. Create a 99-remoteip.conf file inside of conf.modules.d and activated the module by adding the following to the file: 
`LoadModule remoteip_module modules/mod_remoteip.so`

2. Create a remoteip.conf file inside of conf.d and included the following configuration to activate the remoteip module globally (note the use of RemoteIPInternalProxy, that is inclusive of internal private addresses whereas RemoteIPTrustedProxy doesn't include internal private addresses):
```
<IfModule remoteip_module>
    RemoteIPHeader X-Forwarded-For
    RemoteIPInternalProxy <load balancer ip range>
</IfModule>
```

3. Modify my Apache logging to capture %a instead of the default %h so the client's true ip address is captured in the Apache log.

4. Restart Apache to see the changes take effect.

I do appreciate the time you put into your response.

Thanks,
Bryce
